bg-graph/src/main/java/com/dongtronic/diabot/graph/BgGraph.kt

```diff
@@ -0,0 +1,191 @@
+
+    /**
+     * Sets up the chart's axes for BG data.
+     *
+     * @param units The glucose units which are preferred
+     * @return This [BgGraph] instance
+     */
+    fun setupChartAxes(units: String) = apply {
+        val preferredUnit = GlucoseUnit.byName(units) ?: GlucoseUnit.MMOL
+        setupChartAxes(preferredUnit)
+    }
+
+    /**
+     * Sets up the chart's axes for BG data.
+     *
+     * @param preferredUnit The glucose units which are preferred
+     * @return This [BgGraph] instance
+     */
+    fun setupChartAxes(preferredUnit: GlucoseUnit) = apply {
+        requireNonAmbiguous(preferredUnit)
+        // use the preferred unit's axis as y axis group 0.
+        // axis group 0 will be used for creating tick marks on the y-axis for the graph.
+        val mmolGroup = if (preferredUnit == GlucoseUnit.MMOL) 0 else 1
+        val mgdlGroup = if (preferredUnit == GlucoseUnit.MGDL) 0 else 1
+
+        styler.xAxisDecimalPattern = "0.#h"
+        styler.xAxisTickLabelsColor = Color(88, 88, 88)
+
+        styler.isPlotGridVerticalLinesVisible = false
+        styler.plotGridLinesStroke = BasicStroke()
+
+        setYAxisGroupTitle(mgdlGroup, "MG/DL")
+        setYAxisGroupTitle(mmolGroup, "MMOL/L")
+        // always put mmol on the right axis
+        styler.setYAxisGroupPosition(mmolGroup, Styler.YAxisPosition.Right)
+    }
+
+    /**
+     * Adds a [NightscoutDTO]'s BG entries to the chart
+     *
+     * @param nightscout The Nightscout data to add to the chart
+     * @return This [BgGraph] instance
+     */
+    fun addEntries(nightscout: NightscoutDTO) = apply {
+        setupChartAxes(nightscout.units)
+        val readings = nightscout.entries.toList()
+
+        val ranges = mutableMapOf<Color, List<BgEntry>>()
+
+        readings.forEach {
+            val color = getColor(it.glucose.mgdl, nightscout)
+
+            ranges.merge(color, listOf(it)) { oldList: List<BgEntry>, newList: List<BgEntry> ->
+                oldList.plus(newList)
+            }
+        }
+
+        GlucoseUnit.values().forEach { unit ->
+            if (unit == GlucoseUnit.AMBIGUOUS)
+                return@forEach
+
+            // check if these bg values are in the units which are in use by the nightscout instance.
+            //
+            // if the nightscout instance does not have default units (either the settings for this instance have not
+            // been fetched or something went horribly wrong): default to preferred if there's no data in the chart.
+            val preferredUnits = GlucoseUnit.byName(nightscout.units)?.let { it == unit }
+                    ?: seriesMap.isEmpty()
+
+            // don't display mmol/l series on the graph since they're less precise compared to mg/dl
+            val hidden = unit == GlucoseUnit.MMOL
+
+            ranges.forEach { (rangeColor, entries) ->
+                val data = getSeriesData(entries, unit)
+                val color = if (hidden) Color(0, 0, 0, 0) else rangeColor
```

Comment:
Is it possible to add a `filter` step before `forEach` instead of making transparent series?

Reply:
Removing the transparent series will make the mmol axis disappear, so I don't think that's possible unless I'm misunderstanding.

This is what I did to test:
```kotlin
ranges.forEach { (rangeColor, entries) ->
    if (hidden) return@forEach
```

<details>
<summary>Result</summary>

![graph3](https://user-images.githubusercontent.com/32445075/159514895-3c829dfc-3ddd-46ed-86ca-dda1111fb7e6.png)

</details>